### DISCUSSION TITLE
Optimisation

### DISCUSSION THREAD

**Discussion by Lavanya-Bala** [Eomys/pyleecan#358] [Q&A]
Hi,
Before I try optimising my machine, I wanted to try the one done in the webinar (The jupyter notebook mentioned under the Webinar). However, running it gives the following error, **For obj_func : OptiObjective expected, DataKeeper given**
 Am I missing something here ?

**Comment by SebGue** [comment]
The webinar is maybe outdated, i.e. you should use OptiObjective class for the objective (instead of DataKeeper).
You can use it the same way as DataKeeper. The resaon to use OptiObjective is to internally distinguish objectives from normal data to keep.
Best regards, Sebastian

**Reply by Lavanya-Bala** [reply]
Got it, thank you. Further in the optimisation, there is a reference machine used. Could you please tell me the use of this ? My motor has HoleM53 which won't have any issues with changing angles. So, would my reference machine be simply the same ?

**Reply by BonneelP** [reply]
Hello,

I'm not sure if I understood your question (so sorry if my answer is not correct). Which object / part of the tutorial are you referring to ? Normally "the reference machine" should be OptiProblem.simu.machine so it should use your machine with HoleM53 and you can define "simu.machine.rotor.hole[0].W4" as setter for your OptiDesignVar (for the angle). 

Hope it helps,
Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Hi,
Sorry for the confusion. I meant the use of another machine IPMSM_B under Reference simulation in 'How to define a Parameter sweep'

**Comment by BonneelP** [comment]
Yes from the webinar sorry ^^' 
So you can use any machine (as long as pyleecan can simulate it). The idea of "reference machine" is the same principal as for the "reference simulation". When you define your parameter sweep, pyleecan will first run your simulation without the parameter sweep to create a reference. It can be used to speed-up computations (reusing femm files for instance) and get some reference data for the plots (not available yet sadly). 
Then the reference simulation (that contains the reference machine) is copied N times, the modifications from the parameter sweep are applied and each simulation is run separately. 
So it is meant to be your machine not a machine to compare your results to. 

Hope it is clearer now,
Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Yes, that clears it. Thank you for the explanation.

**Reply by Lavanya-Bala** [reply]
The machine I am trying to optimise has magnets'HoleM53' which was created with scripting without any error. But, while optimising it, it throws the error:
**Slot53InterError: ERROR: Slot 53, Can't find Z11 coordinates**

Could you please tell me why this is happening as I can't find any Z11 in the GUI or in the class package of M53?

**Reply by BonneelP** [reply]
Hello,

To draw the holes we compute the coordinates of each points in https://github.com/Eomys/pyleecan/blob/master/pyleecan/Methods/Slot/HoleM53/_comp_point_coordinate.py. My guess is that in the optimization you are trying some value for the angle that doesn't allow to draw the slot. Try to plot your machine with both of your extrema to check that the slot still makes sense with these parameters and if not change them. I think that there may be an option to skip the simulations that raise error in the optimization.

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Further more, I am also getting the following error: 
Traceback (most recent call last):
  File "", line 220, in <module>
    xout3 = simu3.run()
  File "\Methods\Simulation\Simulation\run.py", line 36, in run
    self.var_simu.run()
  File "\Methods\Simulation\VarSimu\run.py", line 56, in run
    post_keeper_postproc_list=self.post_keeper_postproc_list,
  File "\Functions\Simulation\VarSimu\run_single_simu.py", line 55, in run_single_simu
    result = Output(init_dict=Output.as_dict(result))
UnboundLocalError: local variable 'result' referenced before assignment
error: Material properties have not been defined for all regions
Not sure what is going wrong here.

**Reply by BonneelP** [reply]
Hello,
This one looks like a bug in pyleecan. But it seems that you are using an old version of pyleecan since run_single_simu shouldn't exist any more. Which version of pyleecan are you using ? Is it possible for you to update it to the latest version ? 

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
When I try the optimisation (magnet length and slot width of example code) on my machine, it finishes completely but ends with the following error which only seems to appear when I use magnet parameters.  
Traceback (most recent call last):
  File "C:\Users\..\import_class.py", line 8, in import_class
    module = __import__(mod_name + "." + class_name, fromlist=[class_name])
TypeError: can only concatenate str (not "bytes") to str

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/Users/../optimisation.py", line 321, in <module>
    res = load("opti_out.h5")
  File "C:\Users\..\Functions\load.py", line 111, in load
    return init_data(init_dict, file_path)
  File "C:\Users\..\Functions\load.py", line 51, in init_data
    class_obj = import_class("pyleecan.Classes", obj.get("__class__"), "")
  File "C:\Users\..\Load\import_class.py", line 20, in import_class
    "Unknow class name " + class_name + " when loading"
TypeError: can only concatenate str (not "bytes") to str

Could you please let me know if this is a backend bug?

**Reply by SebGue** [reply]
Hello Lavanya-Bala,

do you have the latest pyleecan version installed and the required h5py version, since this is a hdf5 related pyleecan bug.
BR Sebastian

**Reply by Lavanya-Bala** [reply]
Ah yes, thank you.I did not realize that while upgrading to latest version.
